File: AlgoAgentXAPI/app/schemas/funded_backtests.py
from datetime import date
from decimal import Decimal
from typing import Any, Optional
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator

from .backtests import BacktestAdvancedFilters

DAILY_DD_MODES = {"STATIC_INITIAL_BALANCE", "START_OF_DAY_BALANCE", "START_OF_DAY_EQUITY"}
MAX_DD_MODES = {"STATIC_INITIAL_BALANCE", "TRAILING_BALANCE", "TRAILING_EQUITY", "HIGH_WATER_MARK"}
QUALIFYING_MODES = {"ANY_TRADE_DAY", "MIN_PROFIT_DAY"}
KNOWN_CHALLENGE_TYPES = {"ONE_STEP", "TWO_STEP", "INSTANT", "CUSTOM"}
SYSTEM_MAX_RISK = Decimal("0.10")


class FundedPayoutConfig(BaseModel):
    profit_target_pct: Optional[Decimal] = Field(default=None, ge=0)
    profit_target_amount: Optional[Decimal] = Field(default=None, ge=0)
    minimum_payout_amount: Optional[Decimal] = Field(default=None, ge=0)
    minimum_trading_days: Optional[int] = Field(default=None, ge=0)
    minimum_qualifying_profitable_days: Optional[int] = Field(default=None, ge=0)
    qualifying_day_minimum_profit_pct: Optional[Decimal] = Field(default=None, ge=0, lt=1)
    consistency_rule_enabled: bool = False
    consistency_maximum_pct: Optional[Decimal] = Field(default=None, gt=0, le=1)
    consistency_calculation_mode: str = "BEST_DAY_OVER_TOTAL_POSITIVE_PROFIT"
    payout_waiting_calendar_days: Optional[int] = Field(default=None, ge=0)
    payout_profit_split_pct: Optional[Decimal] = Field(default=None, gt=0, le=1)
    stop_simulation_when_payout_ready: bool = False

    @model_validator(mode="after")
    def validate_consistency(self):
        if self.consistency_rule_enabled and self.consistency_maximum_pct is None:
            raise ValueError("consistency_maximum_pct is required when consistency_rule_enabled is true")
        return self


class FundedPhaseInput(BaseModel):
    phase_number: int = Field(..., ge=1)
    phase_name: str = Field(..., min_length=1, max_length=120)
    profit_target_pct: Optional[Decimal] = Field(default=None, ge=0, lt=1)
    daily_drawdown_pct: Decimal = Field(..., gt=0, lt=1)
    daily_drawdown_mode: str
    max_drawdown_pct: Decimal = Field(..., gt=0, lt=1)
    max_drawdown_mode: str
    minimum_trading_days: int = Field(default=0, ge=0)
    minimum_qualifying_day_profit_pct: Optional[Decimal] = Field(default=None, ge=0, lt=1)
    qualifying_day_mode: str = "ANY_TRADE_DAY"
    profit_target_required: bool = True
    reset_balance_after_pass: bool = False
    sequence: int = Field(..., ge=1)

    @field_validator("daily_drawdown_mode")
    @classmethod
    def validate_daily_mode(cls, value):
        value = str(value).upper()
        if value not in DAILY_DD_MODES:
            raise ValueError(f"daily_drawdown_mode must be one of {sorted(DAILY_DD_MODES)}")
        return value

    @field_validator("max_drawdown_mode")
    @classmethod
    def validate_max_mode(cls, value):
        value = str(value).upper()
        if value not in MAX_DD_MODES:
            raise ValueError(f"max_drawdown_mode must be one of {sorted(MAX_DD_MODES)}")
        return value

    @field_validator("qualifying_day_mode")
    @classmethod
    def validate_qualifying_mode(cls, value):
        value = str(value).upper()
        if value not in QUALIFYING_MODES:
            raise ValueError(f"qualifying_day_mode must be one of {sorted(QUALIFYING_MODES)}")
        return value

    @model_validator(mode="after")
    def validate_phase(self):
        if self.profit_target_required and (self.profit_target_pct is None or self.profit_target_pct <= 0):
            raise ValueError("profit_target_pct must be greater than 0 when profit_target_required is true")
        if self.qualifying_day_mode == "MIN_PROFIT_DAY" and self.minimum_qualifying_day_profit_pct is None:
            raise ValueError("minimum_qualifying_day_profit_pct is required for MIN_PROFIT_DAY")
        return self


class FundedRiskTierInput(BaseModel):
    name: str = Field(..., min_length=1, max_length=120)
    sort_order: int = Field(..., ge=1)
    min_account_return_pct: Optional[Decimal] = None
    max_account_return_pct: Optional[Decimal] = None
    risk_percent: Decimal = Field(..., gt=0, le=SYSTEM_MAX_RISK)
    is_active: bool = True

    @model_validator(mode="after")
    def validate_range(self):
        if self.min_account_return_pct is not None and self.max_account_return_pct is not None:
            if self.min_account_return_pct >= self.max_account_return_pct:
                raise ValueError("risk tier min_account_return_pct must be less than max_account_return_pct")
        return self


class FundedProfileBase(BaseModel):
    name: str = Field(..., min_length=1, max_length=255)
    provider_name: Optional[str] = Field(default=None, max_length=255)
    challenge_type: str
    account_size: Decimal = Field(..., gt=0)
    account_currency: str = Field(default="USD", min_length=3, max_length=12)
    is_template: bool = False
    is_active: bool = True
    description: Optional[str] = None
    payout_config: Optional[FundedPayoutConfig] = None
    rules_json: Optional[dict[str, Any]] = None
    phases: list[FundedPhaseInput] = Field(default_factory=list)
    risk_tiers: list[FundedRiskTierInput] = Field(default_factory=list)

    @field_validator("challenge_type")
    @classmethod
    def normalize_challenge_type(cls, value):
        # Extensible string: normalize but do not reject future provider-specific values.
        return str(value).strip().upper()

    @field_validator("account_currency")
    @classmethod
    def normalize_currency(cls, value):
        return str(value).strip().upper()

    @model_validator(mode="after")
    def validate_profile_rules(self):
        phase_numbers = [p.phase_number for p in self.phases]
        sequences = [p.sequence for p in self.phases]
        if len(phase_numbers) != len(set(phase_numbers)):
            raise ValueError("duplicate phase numbers are not allowed")
        if len(sequences) != len(set(sequences)):
            raise ValueError("duplicate phase sequence values are not allowed")
        if self.challenge_type == "ONE_STEP" and (len(self.phases) < 1 or 1 not in phase_numbers):
            raise ValueError("ONE_STEP requires Phase 1")
        if self.challenge_type == "TWO_STEP" and (len(self.phases) < 2 or 1 not in phase_numbers or 2 not in phase_numbers):
            raise ValueError("TWO_STEP requires at least Phase 1 and Phase 2")
        if self.challenge_type == "INSTANT" and self.phases:
            # Instant may have zero challenge phases; custom provider phase constructs should use CUSTOM.
            raise ValueError("INSTANT profiles must not define challenge phases in foundation version")
        self._validate_risk_tiers()
        return self

    def _validate_risk_tiers(self):
        active = sorted((t for t in self.risk_tiers if t.is_active), key=lambda x: x.sort_order)
        orders = [t.sort_order for t in active]
        if len(orders) != len(set(orders)):
            raise ValueError("duplicate risk tier sort_order values are not allowed")
        if active:
            if active[0].min_account_return_pct is not None:
                raise ValueError(f"first active risk tier '{active[0].name}' must have an open lower bound (min_account_return_pct = null)")
            if active[-1].max_account_return_pct is not None:
                raise ValueError(f"last active risk tier '{active[-1].name}' must have an open upper bound (max_account_return_pct = null)")
        for idx, tier in enumerate(active):
            if idx > 0:
                prev = active[idx - 1]
                if prev.max_account_return_pct is None:
                    raise ValueError(f"risk tier '{prev.name}' has an open upper bound before the last tier")
                if tier.min_account_return_pct is None:
                    raise ValueError(f"risk tier '{tier.name}' has an open lower bound after the first tier")
                if tier.min_account_return_pct < prev.max_account_return_pct:
                    raise ValueError(f"risk tiers '{prev.name}' and '{tier.name}' overlap")
                if tier.min_account_return_pct > prev.max_account_return_pct:
                    raise ValueError(f"risk tiers '{prev.name}' and '{tier.name}' leave an uncovered gap")
        return self


class FundedProfileCreate(FundedProfileBase):
    pass


class FundedProfileUpdate(FundedProfileBase):
    pass


class FundedProfileResponse(FundedProfileBase):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    user_id: Optional[UUID] = None


class FundedRunRequest(BaseModel):
    profile_id: UUID
    strategy_id: str
    instrument_id: int
    timeframe: str
    start_date: date
    end_date: date
    runtime_config: Optional[dict[str, Any]] = None
    strategy_preset_id: Optional[str] = None
    advanced_filters: Optional[BacktestAdvancedFilters] = None
    risk_mode: str = "DYNAMIC"
    fixed_risk_pct: Optional[Decimal] = Field(default=None, gt=0, le=SYSTEM_MAX_RISK)
    safety_buffer_pct: Decimal = Field(default=Decimal("0"), ge=0, lt=1)
    configured_max_risk_pct: Optional[Decimal] = Field(default=None, gt=0, le=SYSTEM_MAX_RISK)

    @field_validator("risk_mode")
    @classmethod
    def normalize_risk_mode(cls, value):
        value = str(value or "DYNAMIC").strip().upper()
        if value not in {"DYNAMIC", "FIXED"}:
            raise ValueError("risk_mode must be DYNAMIC or FIXED")
        return value

    @model_validator(mode="after")
    def validate_dates(self):
        if self.end_date < self.start_date:
            raise ValueError("end_date must be on or after start_date")
        if self.risk_mode == "FIXED" and self.fixed_risk_pct is None:
            raise ValueError("fixed_risk_pct is required when risk_mode is FIXED")
        return self

File: AlgoAgentXApp/app/lib/api/funded-backtests.ts
import axiosInstance from "@/lib/axios";
import { HEAVY_API_TIMEOUT_MS, POLLING_API_TIMEOUT_MS } from "@/lib/api-timeouts";
import type { AdvancedBacktestFilters } from "@/lib/api/backtests";

type ApiEnvelope<T> = { success?: boolean; data?: T; message?: string };
const unwrap = <T,>(payload: ApiEnvelope<T> | T): T => {
  if (payload && typeof payload === "object" && "success" in (payload as Record<string, unknown>)) {
    return ((payload as ApiEnvelope<T>).data ?? null) as T;
  }
  return payload as T;
};

export type ChallengeType = "ONE_STEP" | "TWO_STEP" | "INSTANT" | "CUSTOM" | string;
export type DailyDdMode = "STATIC_INITIAL_BALANCE" | "START_OF_DAY_BALANCE" | "START_OF_DAY_EQUITY";
export type MaxDdMode = "STATIC_INITIAL_BALANCE" | "TRAILING_BALANCE" | "TRAILING_EQUITY" | "HIGH_WATER_MARK";
export type QualifyingDayMode = "ANY_TRADE_DAY" | "MIN_PROFIT_DAY";
export type FundedRiskMode = "FIXED" | "DYNAMIC";

export interface FundedPayoutRules {
  profit_target_pct?: number | null;
  profit_target_amount?: number | null;
  minimum_payout_amount?: number | null;
  minimum_trading_days?: number | null;
  minimum_qualifying_profitable_days?: number | null;
  qualifying_day_minimum_profit_pct?: number | null;
  consistency_rule_enabled?: boolean;
  consistency_maximum_pct?: number | null;
  consistency_calculation_mode?: string;
  payout_waiting_calendar_days?: number | null;
  payout_profit_split_pct?: number | null;
  stop_simulation_when_payout_ready?: boolean;
}

export interface FundedAccountPhase {
  phase_number: number;
  phase_name: string;
  profit_target_pct?: number | null;
  daily_drawdown_pct: number;
  daily_drawdown_mode: DailyDdMode | string;
  max_drawdown_pct: number;
  max_drawdown_mode: MaxDdMode | string;
  minimum_trading_days: number;
  minimum_qualifying_day_profit_pct?: number | null;
  qualifying_day_mode: QualifyingDayMode | string;
  profit_target_required: boolean;
  reset_balance_after_pass: boolean;
  sequence: number;
}

export interface FundedRiskTier {
  name: string;
  sort_order: number;
  min_account_return_pct?: number | null;
  max_account_return_pct?: number | null;
  risk_percent: number;
  is_active: boolean;
}

export interface FundedAccountProfile {
  id: string;
  user_id?: string | null;
  name: string;
  provider_name?: string | null;
  challenge_type: ChallengeType;
  account_size: number;
  account_currency: string;
  is_template: boolean;
  is_active: boolean;
  description?: string | null;
  payout_config?: FundedPayoutRules | null;
  rules_json?: Record<string, unknown> | null;
  phases: FundedAccountPhase[];
  risk_tiers: FundedRiskTier[];
  created_at?: string;
  updated_at?: string;
}

export type FundedProfilePayload = Omit<FundedAccountProfile, "id" | "user_id" | "created_at" | "updated_at">;

export interface FundedRunRequest {
  profile_id: string;
  strategy_id: string;
  instrument_id: number;
  timeframe: string;
  start_date: string;
  end_date: string;
  runtime_config?: Record<string, unknown>;
  strategy_preset_id?: string | null;
  advanced_filters?: AdvancedBacktestFilters;
  risk_mode: FundedRiskMode;
  fixed_risk_pct?: number | null;
  safety_buffer_pct?: number;
  configured_max_risk_pct?: number | null;
}

export interface FundedPreviewResponse {
  valid: boolean;
  rule_engine_version?: string;
  profile_id: string;
  strategy_id: string;
  instrument_id: number;
  account_name: string;
  challenge_type: string;
  initial_capital: number;
  risk_mode: FundedRiskMode;
  fixed_risk_pct?: number | null;
  safety_buffer_pct?: number;
  market_data?: { candles?: number; from?: string; to?: string };
  snapshot_preview?: Record<string, any>;
  warnings?: string[];
  credit_policy?: string;
}

export interface FundedRunResponse {
  funded_backtest_id: string;
  status: string;
  current_phase?: number | null;
  final_balance?: number | null;
  final_equity?: number | null;
  trades_processed?: number;
  summary?: Record<string, any>;
  job_id?: string;
}

export interface FundedRunStatus {
  id: string;
  status: string;
  current_phase?: number | null;
  trades_processed?: number;
  current_balance?: number | null;
  current_equity?: number | null;
  target_progress?: Record<string, any> | number | null;
  trading_days?: number;
  qualifying_days?: number;
  failure_reason?: string | null;
  payout?: Record<string, any> | null;
  technical_error?: string | null;
}

export interface FundedRunDetail extends FundedRunStatus {
  profile_id?: string;
  strategy_id?: string;
  instrument_id?: number;
  timeframe?: string;
  start_date?: string;
  end_date?: string;
  initial_capital?: number;
  final_balance?: number | null;
  final_equity?: number | null;
  passed_at?: string | null;
  failed_at?: string | null;
  payout_eligible_at?: string | null;
  calendar_days?: number;
  summary?: Record<string, any>;
  phase_summaries?: Array<Record<string, any>>;
  account_snapshot?: Record<string, any>;
  runtime_snapshot?: Record<string, any>;
  risk_plan_snapshot?: Record<string, any>;
}

export const fundedBacktestsApi = {
  async listProfiles(): Promise<FundedAccountProfile[]> {
    const response = await axiosInstance.get<ApiEnvelope<FundedAccountProfile[]> | FundedAccountProfile[]>("/api/v1/funded-backtests/profiles");
    const data = unwrap(response.data);
    return Array.isArray(data) ? data : [];
  },
  async getProfile(id: string): Promise<FundedAccountProfile> {
    const response = await axiosInstance.get<ApiEnvelope<FundedAccountProfile> | FundedAccountProfile>(`/api/v1/funded-backtests/profiles/${id}`);
    return unwrap(response.data);
  },
  async createProfile(payload: FundedProfilePayload): Promise<FundedAccountProfile> {
    const response = await axiosInstance.post<ApiEnvelope<FundedAccountProfile> | FundedAccountProfile>("/api/v1/funded-backtests/profiles", payload);
    return unwrap(response.data);
  },
  async updateProfile(id: string, payload: FundedProfilePayload): Promise<FundedAccountProfile> {
    const response = await axiosInstance.put<ApiEnvelope<FundedAccountProfile> | FundedAccountProfile>(`/api/v1/funded-backtests/profiles/${id}`, payload);
    return unwrap(response.data);
  },
  async deactivateProfile(id: string): Promise<{ id: string; is_active: boolean }> {
    const response = await axiosInstance.delete<ApiEnvelope<{ id: string; is_active: boolean }> | { id: string; is_active: boolean }>(`/api/v1/funded-backtests/profiles/${id}`);
    return unwrap(response.data);
  },
  async preview(payload: FundedRunRequest): Promise<FundedPreviewResponse> {
    const response = await axiosInstance.post<ApiEnvelope<FundedPreviewResponse> | FundedPreviewResponse>("/api/v1/funded-backtests/preview", payload, { timeout: HEAVY_API_TIMEOUT_MS });
    return unwrap(response.data);
  },
  async run(payload: FundedRunRequest): Promise<FundedRunResponse> {
    const response = await axiosInstance.post<ApiEnvelope<FundedRunResponse> | FundedRunResponse>("/api/v1/funded-backtests/run", payload, { timeout: HEAVY_API_TIMEOUT_MS });
    return unwrap(response.data);
  },
  async getStatus(id: string): Promise<FundedRunStatus> {
    const response = await axiosInstance.get<ApiEnvelope<FundedRunStatus> | FundedRunStatus>(`/api/v1/funded-backtests/${id}/status`, { timeout: POLLING_API_TIMEOUT_MS });
    return unwrap(response.data);
  },
  async getRun(id: string): Promise<FundedRunDetail> {
    const response = await axiosInstance.get<ApiEnvelope<FundedRunDetail> | FundedRunDetail>(`/api/v1/funded-backtests/${id}`, { timeout: HEAVY_API_TIMEOUT_MS });
    return unwrap(response.data);
  },
};
